WebGLRenderer: Add outputBufferType and setEffects()

mrdoob opened the pull request:
Related issue: #15840 #18322 #18846 #32433 #32449 

**Description**

Adds internal HDR rendering and simplified post-processing API.

**Changes**

* `outputBufferType` option for HalfFloatType rendering
* `setEffects()` method for post-processing without EffectComposer
* `isRenderPass` flag for TAARenderPass/SSAARenderPass support

```js
const renderer = new WebGLRenderer({ outputBufferType: HalfFloatType });
renderer.setEffects( [ taaPass, bloomPass ] );
renderer.render( scene, camera );
```

| Before | After |
| - | - |
| <img width="628" height="531" alt="Screenshot 2025-12-02 at 20 07 24" src="https://github.com/user-attachments/assets/da698471-17e9-420e-ab5c-e25865ef1fe1" /> | <img width="628" height="531" alt="Screenshot 2025-12-02 at 20 07 31" src="https://github.com/user-attachments/assets/9febc8de-f77c-40fb-a668-115d26b4133d" /> |
| <img width="628" height="531" alt="Screenshot 2025-12-02 at 20 08 21" src="https://github.com/user-attachments/assets/b8c7f865-0ab2-441b-aac5-e1e9b084e842" /> | <img width="628" height="531" alt="Screenshot 2025-12-02 at 20 08 24" src="https://github.com/user-attachments/assets/f60a19db-ed3e-4a7f-9bab-9324440b1bd3" /> |

The watch glass HDR is fixed but the text becomes less sharp. I still don't understand why...

| Before | After |
| - | - |
| <img width="645" height="658" alt="Screenshot 2025-12-02 at 20 22 52" src="https://github.com/user-attachments/assets/e0cb8fd6-4b01-4ec6-b6f5-9b272aac4e5b" /> | <img width="645" height="658" alt="Screenshot 2025-12-02 at 20 22 59" src="https://github.com/user-attachments/assets/a3bb9a00-54f4-47bd-8f10-48a0f7f3bc5b" /> |

For testing:
https://raw.githack.com/mrdoob/three.js/733f7abc2378874b0c386010b675b7a43f5c6aa5/examples/index.html

Mugen87 commented:
BTW: An optional setup with `WebGLRenderer` is via `EffectComposer`. A setup of a render pass + output psass should produce similar results.

mrdoob committed:
WebGLRenderer: Add colorBufferType and setEffects().

mrdoob committed:
Updated watch and ldraw examples.

mrdoob commented:
Refactored the PR:

* Added `colorBufferType` to `WebGLRenderer` (`UnsignedByteType` by default)
* Also added `.setEffects()` which tries to simplify `EffectComposer`.

@gkjohnson Changing `colorBufferType` to solve the clamped transparency issues feels much more easier to understand to me than messing with premultiplied alpha. Both options have their downsides but this one seems better as it also aligns with `WebGPURenderer`.

mrdoob committed:
WebGLRenderer: Revert transmissionRenderTarget type change.

mrdoob committed:
Clean up.

mrdoob committed:
Renamed colorBufferType to outputBufferType.

mrdoob committed:
Added effects support to WebXR.

mrdoob committed:
Updated screenshot.

mrdoob committed:
We no longer need to resize the passes manually.

mrdoob committed:
Fixed feedback loop in webgl_watch.

Mugen87 commented:
> Instead of .outputType, how about calling it .canvasFormat?

Since we still assign a type (like `HalfFloatType`), maybe `canvasType`?

mrdoob commented:
> > Instead of .outputType, how about calling it .canvasFormat?
> 
> Since we still assign a type (like `HalfFloatType`), maybe `canvasType`?

Or `.canvasFormatType`.

mrdoob committed:
Clean up.

mrdoob committed:
Adjusted bloom strength in webgl_watch.

mrdoob committed:
WebGLRenderer: Warn user when using redundant OutputPass in setEffects()

mrdoob committed:
Renamed WebGLOutputBuffer to WebGLOutput.

mrdoob committed:
Simplified WebGLOutput.

mrdoob committed:
Clean up.

mrdoob committed:
Simplified watch example.

mrdoob committed:
WebGLRenderer: Only allow effects when outputBufferType isn't UnsignedByteType.

mrdoob committed:
Clean up.
